client/x/evmstaking/keeper/withdraw.go

```diff
@@ -239,92 +168,148 @@ func (k Keeper) ExpectedRewardWithdrawals(ctx context.Context) ([]RewardWithdraw
+		return err
+	}
 
-		valAccAddr := sdk.AccAddress(valAddr).String()
+	delRewards, err := k.distributionKeeper.CalculateDelegationRewards(ctx, validator, delegation, endingPeriod)
+	if err != nil {
+		return err
+	}
 
-		// Withdraw delegation rewards.
-		delAddr := sdk.MustAccAddressFromBech32(rewardWithdrawals[i].DelegatorAddress)
-		delRewards, err := k.distributionKeeper.WithdrawDelegationRewards(ctx, delAddr, valAddr)
+	// if it is self-delegation, add commission
+	if delegation.DelegatorAddress == valAccAddr.String() {
+		// Get validator commissions.
+		valCommission, err := k.distributionKeeper.GetValidatorAccumulatedCommission(ctx, valAddr)
 		if err != nil {
 			return err
 		}
 
-		// Withdraw commission if it is a self delegation.
-		if rewardWithdrawals[i].DelegatorAddress == valAccAddr {
-			commissionRewards, err := k.distributionKeeper.WithdrawValidatorCommission(ctx, valAddr)
-			if errors.Is(err, dtypes.ErrNoValidatorCommission) {
-				log.Debug(
-					ctx, "No validator commission",
-					"validator_addr", rewardWithdrawals[i].ValidatorAddress,
-					"validator_account_addr", valAccAddr,
-				)
-			} else if err != nil {
-				return err
-			} else {
-				delRewards = delRewards.Add(commissionRewards...)
-			}
-		}
+		delRewards = delRewards.Add(valCommission.Commission...)
+	}
 
-		if rewardWithdrawals[i].RemainedBalance > 0 {
-			rewardCoins := sdk.NewCoin(sdk.DefaultBondDenom, math.NewInt(int64(rewardWithdrawals[i].RemainedBalance)))
-			delRewards = delRewards.Add(rewardCoins)
-		}
+	addr, err := sdk.AccAddressFromBech32(delegation.DelegatorAddress)
+	if err != nil {
+		return errors.Wrap(err, "convert acc address from bech32 address")
+	}
 
-		delRewardUint64 := delRewards.AmountOf(sdk.DefaultBondDenom).Uint64()
-		log.Debug(
-			ctx, "Withdraw delegator rewards",
-			"validator_addr", rewardWithdrawals[i].ValidatorAddress,
-			"validator_account_addr", valAccAddr,
-			"delegator_addr", rewardWithdrawals[i].DelegatorAddress,
-			"amount_calculate", rewardWithdrawals[i].WithdrawalEntry.Amount,
-			"amount_deposited", rewardWithdrawals[i].RemainedBalance,
-			"amount_total_withdrawal", delRewardUint64,
-		)
+	delRewardsTruncated, _ := delRewards.TruncateDecimal()
+	unclaimedReward := delRewardsTruncated.AmountOf(sdk.DefaultBondDenom).Uint64()
+	claimedReward := k.bankKeeper.SpendableCoin(ctx, addr, sdk.DefaultBondDenom).Amount.Uint64()
+	totalReward := unclaimedReward + claimedReward
 
-		// Burn tokens from the delegator
-		err = k.bankKeeper.SendCoinsFromAccountToModule(ctx, delAddr, types.ModuleName, delRewards)
-		if err != nil {
-			return err
-		}
-		err = k.bankKeeper.BurnCoins(ctx, types.ModuleName, delRewards)
-		if err != nil {
-			return err
+	// if total reward is greater than or equal to min reward withdrawal amount, enqueue the reward withdrawal
+	if totalReward >= minRewardWithdrawalAmount {
+		if err := k.EnqueueRewardWithdrawal(ctx, delegation.DelegatorAddress, valAddr.String(), claimedReward); err != nil {
```

Comment:
This switches from calling `EnqueueEligibleRewardWithdrawal` after calculating all eligible reward withdrawals (all iterations of delegations) to calling `EnqueueRewardWithdrawal` in each iteration of the delegations. 

Wondering if this breaks any assumption in each iteration's calculation (e.g. we might expect the validator's balance to remain static throughout the loop iterations), or if no assumptions are broken with this new place for logic execution.

Reply:
Basically, in CL side, there is no asset unbonded (so transferrable) as far as I know, because staking related actions are started from EL side and unbonded assets are burnt when undelegated. The only case that CL account has asset in bank keeper is that when the delegator makes another delegation under the same validator. (by settling the previous rewards and creating a new starting info for reward).

If you have any cases which could breaks any assumptions in your mind, please lmk. Or any other suggestions on handling this would be great.

Reply:
Thanks, looks good for now.